File: frontend/src/lib/services/TechService.ts
import techjson from '$lib/ssr/techs.json';
import {
	GetTechsResponseSchema,
	type GetTechsResponseJson,
	type TechDefense,
	type TechHull,
	type TechHullComponent,
	type TechPlanetary,
	type TechPlanetaryScanner,
	type TechTerraform
} from '$lib/types/cs-proto';
import type { TechLike, TechStore } from '$lib/types/Tech';
import { fromJson, type UnknownField } from '@bufbuild/protobuf';
import { kebabCase } from 'lodash-es';

export class TechService implements TechStore {
	$typeName: 'craig_stars.v1.GetTechsResponse';
	$unknown?: UnknownField[] | undefined;

	techs: TechLike[] = [];
	planetaryScanners: TechPlanetaryScanner[] = [];
	terraforms: TechTerraform[] = [];
	defenses: TechDefense[] = [];
	planetaries: TechPlanetary[] = [];
	hullComponents: TechHullComponent[] = [];
	hulls: TechHull[] = [];

	techsByName: Map<string, TechLike> = new Map();
	hullsByName: Map<string, TechHull> = new Map();
	hullComponentsByName: Map<string, TechHullComponent> = new Map();

	constructor(store?: TechStore) {
		this.$typeName = 'craig_stars.v1.GetTechsResponse';
		store = store ?? fromJson(GetTechsResponseSchema, techjson as unknown as GetTechsResponseJson);
		this.buildMaps(store);
	}

	private buildMaps(store: TechStore) {
		this.planetaryScanners = store.planetaryScanners ?? [];
		this.terraforms = store.terraforms ?? [];
		this.defenses = store.defenses ?? [];
		this.planetaries = store.planetaries ?? [];
		this.hullComponents = store.hullComponents ?? [];
		this.hulls = store.hulls ?? [];

		this.techs = [];
		this.techs = this.techs.concat(this.hullComponents);
		this.techs = this.techs.concat(this.planetaryScanners);
		this.techs = this.techs.concat(this.defenses);
		this.techs = this.techs.concat(this.planetaries);
		this.techs = this.techs.concat(this.hulls);
		this.techs = this.techs.concat(this.terraforms);

		this.techsByName = new Map(this.techs.map((t) => [kebabCase(t.tech?.name), t]));
		this.hullComponentsByName = new Map(
			this.hullComponents.map((t) => [kebabCase(t.tech?.name), t])
		);
		this.hullsByName = new Map(this.hulls.map((t) => [kebabCase(t.tech?.name), t]));
	}

	getTech(name: string): TechLike | undefined {
		return this.techsByName.get(kebabCase(name));
	}

	getHull(name: string): TechHull | undefined {
		return this.hullsByName.get(kebabCase(name));
	}

	getHullComponent(name: string): TechHullComponent | undefined {
		return this.hullComponentsByName.get(kebabCase(name));
	}
}

File: server/techservice.go
//go:build !wasi && !wasm

package server

import (
	"context"
	"fmt"

	"connectrpc.com/connect"
	"github.com/sirgwain/craig-stars/cs"
	"github.com/sirgwain/craig-stars/proto/converter"
	craig_starsv1 "github.com/sirgwain/craig-stars/proto/gen/craig_stars/v1"
	"github.com/sirgwain/craig-stars/proto/gen/craig_stars/v1/craig_starsv1connect"
)

// NewTechServiceHandler constructs the TechService handler.
func NewTechServiceHandler() craig_starsv1connect.TechServiceHandler {
	return &techService{}
}

type techService struct{}

// GetTech returns a single Tech by name, or NOT_FOUND if missing.
func (s *techService) GetTech(ctx context.Context, req *connect.Request[craig_starsv1.GetTechRequest]) (*connect.Response[craig_starsv1.GetTechResponse], error) {
	if req.Msg.Name == "" {
		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("name is required"))
	}
	tech := cs.StaticTechStore.GetTech(req.Msg.GetName())
	if tech == nil {
		return nil, connect.NewError(connect.CodeNotFound, fmt.Errorf("tech not found"))
	}

	resp := &craig_starsv1.GetTechResponse{}
	switch t := tech.(type) {
	case *cs.TechPlanetaryScanner:
		resp.Tech = &craig_starsv1.GetTechResponse_TechPlanetaryScanner{
			TechPlanetaryScanner: converter.C.ConvertCSTechPlanetaryScanner(t),
		}
	case *cs.TechTerraform:
		resp.Tech = &craig_starsv1.GetTechResponse_TechTerraform{
			TechTerraform: converter.C.ConvertCSTechTerraform(t),
		}
	case *cs.TechDefense:
		resp.Tech = &craig_starsv1.GetTechResponse_TechDefense{
			TechDefense: converter.C.ConvertCSTechDefense(t),
		}
	case *cs.TechPlanetary:
		resp.Tech = &craig_starsv1.GetTechResponse_TechPlanetary{
			TechPlanetary: converter.C.ConvertCSTechPlanetary(t),
		}
	case *cs.TechHullComponent:
		resp.Tech = &craig_starsv1.GetTechResponse_TechHullComponent{
			TechHullComponent: converter.C.ConvertCSTechHullComponentP(t),
		}
	case *cs.TechHull:
		resp.Tech = &craig_starsv1.GetTechResponse_TechHull{
			TechHull: converter.C.ConvertCSTechHull(t),
		}
	default:
		return nil, connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("unsupported tech type"))
	}

	return connect.NewResponse(resp), nil
}

// GetTechs returns the full TechStore (StaticTechStore).
func (s *techService) GetTechs(ctx context.Context, req *connect.Request[craig_starsv1.GetTechsRequest]) (*connect.Response[craig_starsv1.GetTechsResponse], error) {
	techs := cs.StaticTechStore
	return connect.NewResponse(&craig_starsv1.GetTechsResponse{
		PlanetaryScanners: converter.C.ConvertCSTechPlanetaryScanners(techs.PlanetaryScanners),
		Terraforms:        converter.C.ConvertCSTechTerraforms(techs.Terraforms),
		Defenses:          converter.C.ConvertCSTechDefenses(techs.Defenses),
		Planetaries:       converter.C.ConvertCSTechPlanetaries(techs.Planetaries),
		HullComponents:    converter.C.ConvertCSTechHullComponents(techs.HullComponents),
		Hulls:             converter.C.ConvertCSTechHulls(techs.Hulls),
	}), nil
}
